### PR TITLE
simulateSam - Error and Coverage

### PR DESCRIPTION
**Pull request by rimangan**
# Description
This PR introduces two new features to the cmd simulateSam. 

### Coverage
Previously, the user could input a desired number of reads to be generated by the program. Here, I introduce 'coverage', which allows the user to specify a desired expected coverage value instead. The program then automatically calculates the number of reads to generate based on the genome size, desired coverage, and bases per read.
New usage, to generate 30x sequencing coverage, looks like this:
~/go/bin/simulateSam -coverage 30 hg38.fa output/hg38.30xCoverage.bam

### Error Rate
This option allows the user to specify a sequencing/PCR error rate in simulating sam reads. If -flatErrorRate is set to
0.01, for example, one percent of A's will appear as either a C, G, or T in the simulated read. Error rates are flat, in that an A has an equal probability of mutating to a C, G, or T. Note that I plan later to implement uneven error rates to simulate the case of ancient DNA samples with high amounts of cytosine deamination.

## Relevant Links

- [Display Text](https://www.vertgenlab.org/)

### Checklist before requesting a review

- [x] I performed a self-review of my code
- [x] If it this a core feature, I have added thorough tests
- [x] The command `go fmt` was used on all files included

### Testing

None

### Screenshots & Media

None

### Edge cases / Breaking Changes / Known Issues

None

### CODE REVIEW

**Review by edotau — APPROVED** [review]
lgtm!

**Comment by rimangan** [comment]
Marking as "Do Not Merge" as Craig came up with a valuable performance idea that I want to implement. More info on that once I can implement.

**Comment by rimangan** [comment]
On a previous commit, I simulated error at an error rate epsilon by marching through the read sequence, generating a random number, and checking if the random number was less than epsilon. While this produces the intended result, it runs in O(n) time, checking each base. For generating large whole-genome synthetic libraries, this will be very unnecessarily intensive. 
In the new commit, I first generate the number of errors, pick that many bases without replacement, and mutate those bases to a new base. Now, instead of a check at every base, performance scales with the error rate, specifically the expected value of errors per read, which in most real world settings will only be one or two base pairs (instead of 150bp).

We treat errors like a binomial process, where for 150bp in the read, each base has an equal probability of mutating, with rate epsilon. Therefore, we can sample a binomial distribution to get the number of errors.

The trick to this whole speedup is that there is an efficient algorithm, Walker's Alias method, which allows us to generate binomially-distributed random variates in O(1) time. Full details here: https://en.wikipedia.org/wiki/Alias_method.
I've implemented this algorithm in the new file randBinomial.go, in the numbers package.

I'm happy to talk through the algorithm specifics in code review. Here are some snapshots of its performance. The top histogram is the true binomial distribution for given parameters, and the bottom histogram is the empirical histogram from 1000 variates.
image

File: simulate/sam.go [hunk 1]
```diff
@@ -53,53 +53,71 @@ func IlluminaPairedSam(refName string, ref []dna.Base, numPairs, readLen, avgFra
...
 	}
 	for realSeqEndIdx = end; realSeqEndIdx > len(ref); realSeqEndIdx-- {
-		s.Seq[len(s.Seq)-(1+(end-realSeqEndIdx))] = dna.Base(numbers.RandIntInRange(0, 4))
+		currSam.Seq[len(currSam.Seq)-(1+(end-realSeqEndIdx))] = dna.Base(numbers.RandIntInRange(0, 4))
+	}
+	copy(currSam.Seq[realSeqStartIdx-start:len(currSam.Seq)-(end-realSeqEndIdx)], ref[realSeqStartIdx:realSeqEndIdx])
+
+	// now we will simulate sequencing/PCR error with a flat error rate
+	if flatErrorRate > 0 {
+		numFlatErrors := numbers.RandBinomial(alias) // sample a binomial distribution to get the number of sequencing errors
```
Comment (craiglowe) [inline simulate/sam.go:91]: It seems like what is inside of this if-statement could be a function call where you pass a sam to have errors added to it and get it back

**Review by craiglowe — COMMENTED** [review]
Great work!  I had a few minor comments

File: cmd/simulateSam/simulateSam.go [hunk 2]
```diff
@@ -75,6 +79,12 @@ func getReadsPerContig(ref []fasta.Fasta, numReads int) []int {
 		cdf[i] = cdf[i-1] + contigWeights[i]
 	}
 
+	//if the user specified coverage, we will set numReads.
+	// readLen*2 for paired end gets bases per molecule.
+	if coverage > 0 {
+		numReads = int(coverage * float64(totalLen) / float64(readLen*2))
```
[the remark this answers is not in the document]
Comment (rimangan) [reply cmd/simulateSam/simulateSam.go:85]: I would agree but the problem is that we need "totalLen" to calculate numReads from coverage, and we don't calculate totalLen until we're inside getReadsPerContig

**Review by craiglowe — COMMENTED** [review]
Thank you!  Just one more suggestion for a small change to simplify a function call.

File: simulate/sam.go [hunk 3]
```diff
@@ -143,3 +148,21 @@ func addPairedFlags(f, r *sam.Sam) {
 		}
 	}
 }
+
+func sequencingError(currSam sam.Sam, start int, end int, alias numbers.BinomialAlias) sam.Sam {
```
[the remark this answers is not in the document]
Comment (rimangan) [reply simulate/sam.go]: in next commit

**Review by craiglowe — APPROVED** [review]
Wonderful!

**Review by edotau — APPROVED** [review]
